cardano-config/src/Cardano/Node/Configuration/File.hs

```diff
@@ -0,0 +1,135 @@
+{-# LANGUAGE GADTs #-}
+
+-- | The representation of the configuration file
+module Cardano.Node.Configuration.File
+  ( -- * Configuration file
+    NodeConfigurationFromFile
+  , NodeConfigurationFromFileF (..)
+  , parseConfigurationFiles
+
+    -- * Specific components configurations
+  , StorageConfiguration (..)
+  , ConsensusConfiguration (..)
+  , ProtocolConfiguration (..)
+  , NetworkConfiguration (..)
+  , LocalConnectionsConfig (..)
+  , TracingConfiguration (..)
+  , TestingConfiguration (..)
+  ) where
+
+import Cardano.Node.Configuration.File.Consensus
+import Cardano.Node.Configuration.File.Network
+import Cardano.Node.Configuration.File.Protocol
+import Cardano.Node.Configuration.File.Storage
+import Cardano.Node.Configuration.File.Testing
+import Cardano.Node.Configuration.File.Tracing
+import Control.Applicative ((<|>))
+import Control.Monad.Except
+import Control.Monad.Trans (lift)
+import Data.Bifunctor (bimap)
+import qualified Data.ByteString as BS
+import Data.Functor.Identity (Identity (..))
+import Data.String (fromString)
+import Data.Yaml
+import GHC.Generics (Generic)
+import System.Directory (doesFileExist)
+import System.FilePath (takeDirectory, (</>))
+
+-- | The configuration from the files, parsed with 'parseConfigurationFiles'
+type NodeConfigurationFromFile = NodeConfigurationFromFileF Identity
+
+-- | The configuration from the files, initially maybe pointing to sub-files and
+-- finally fully parsed.
+data NodeConfigurationFromFileF f
+  = NodeConfigurationFromFileV1
+  { storageConfiguration :: f (StorageConfiguration Maybe)
+  , consensusConfiguration :: f ConsensusConfiguration
+  , protocolConfiguration :: f (ProtocolConfiguration Maybe)
+  , networkConfiguration :: f NetworkConfiguration
+  , localConnectionsConfig :: LocalConnectionsConfig
+  , tracingConfiguration :: f TracingConfiguration
+  , testingConfiguration :: TestingConfiguration
+  }
+  deriving Generic
+
+deriving instance Show (NodeConfigurationFromFileF (Either FilePath))
+deriving instance Show (NodeConfigurationFromFileF Identity)
+
+-- | Either an action to parse another file, or directly a value
+data SubFileOrValue a
+  = SubFile (IO (Either String a))
+  | AValue a
+
+-- | If the consulted key is a filepath, then prepare an action to parse that
+-- other file.
+subFileParser ::
+  FilePath ->
+  String ->
+  (Value -> Parser a) ->
+  Value ->
+  Parser (SubFileOrValue a)
+subFileParser root sectionName p v =
+  withObject
+    ("Config file: " <> sectionName)
+    ( ( \v' -> do
+          fn <- v' .: fromString sectionName
+          pure $
+            SubFile $ do
+              exists <- doesFileExist $ root </> fn
+              if exists
+                then
+                  parseEither p
+                    . either (\err -> error $ "Invalid JSON: " <> show err) id
+                    . decodeEither'
+                    <$> BS.readFile (root </> fn)
```

Comment:
So this expression can be:
- an IO exception
- a bottom from `error` wrapped in `Left` (in `IO`)
- a parser failure as `Left` (in `IO`)
- a successful parse as `Right` (in `IO`)



What's the advantage of having three different levels of errors here? Why not just work here in `IO a`?

All those different kinds of errors make this function just harder to reason about without much benefit IMO.

Shouldn't the second and third type of errors be short-circuited to the topmost parser instead i.e. this one  `Parser (SubFileOrValue a)`?